### PR TITLE
Add support for key ID in tokens

### PR DESCRIPTION
**Pull request by paulgb**
This adds support for a concept of "key IDs" in token authentication.

This is a feature of the "core" library only, and is not exposed by this PR in either the native or Cloudflare server.

Key IDs are used by consumers of the core library by calling `with_key_id(key_id)`
on an `Authenticator`, which returns a new `Authenticator` using that key ID.

An `Authenticator` with a key ID has its behavior changed:

- When generating keys, it prepends `{key_id}.`.
- When validating tokens, it expects to find `{key_id}.` as a prefix, which is stripped
  before validating the token as usual.

### CODE REVIEW

File: crates/y-sweet-core/src/auth.rs [hunk 1]
```diff
@@ -165,24 +129,81 @@ fn hash(bytes: &[u8]) -> Vec<u8> {
...
+        if expected_token != auth_req.token {
+            Err(AuthError::InvalidSignature)
+        } else if auth_req.payload.expiration_millis.unwrap_or(u64::MAX) < current_time {
+            Err(AuthError::Expired)
+        } else {
+            Ok(auth_req.payload)
+        }
+    }
+
+    pub fn with_key_id(self, key_id: String) -> Self {
```
Comment (pretentious7) [inline crates/y-sweet-core/src/auth.rs:194]: does the key id itself not have to be url safe base 64?

File: crates/y-sweet-core/src/auth.rs [hunk 2]
```diff
@@ -165,24 +129,81 @@ fn hash(bytes: &[u8]) -> Vec<u8> {
...
+        } else {
+            if self.key_id.is_some() {
+                return Err(AuthError::KeyMismatch);
+            }
+
+            &token
+        };
+
+        let auth_req: AuthenticatedRequest =
+            bincode_decode(&b64_decode(token)?).map_err(|_| AuthError::InvalidToken)?;
```
Comment (pretentious7) [inline crates/y-sweet-core/src/auth.rs:177]: or instead of map err

File: crates/y-sweet-core/src/auth.rs (hunk 1 above)
Comment (pretentious7) [reply crates/y-sweet-core/src/auth.rs:194]: ah, it doesn't since it's in the header.

are there any limitations there? or can that be any utf-8

Comment (paulgb) [reply crates/y-sweet-core/src/auth.rs:194]: It has a restricted alphabet, I guess it doesn’t hurt to check for that. Maybe I’ll restrict to url-safe base64

**Pull request merged**